man/lmtpd.8
.\" -*- nroff -*-
.TH LMTPD 8 "Project Cyrus" CMU
.\" 
.\" Copyright (c) 1998-2000 Carnegie Mellon University.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\"
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer. 
.\"
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in
.\"    the documentation and/or other materials provided with the
.\"    distribution.
.\"
.\" 3. The name "Carnegie Mellon University" must not be used to
.\"    endorse or promote products derived from this software without
.\"    prior written permission. For permission or any other legal
.\"    details, please contact  
.\"      Office of Technology Transfer
.\"      Carnegie Mellon University
.\"      5000 Forbes Avenue
.\"      Pittsburgh, PA  15213-3890
.\"      (412) 268-4387, fax: (412) 268-7395
.\"      tech-transfer@andrew.cmu.edu
.\"
.\" 4. Redistributions of any form whatsoever must retain the following
.\"    acknowledgment:
.\"    "This product includes software developed by Computing Services
.\"     at Carnegie Mellon University (http://www.cmu.edu/computing/)."
.\"
.\" CARNEGIE MELLON UNIVERSITY DISCLAIMS ALL WARRANTIES WITH REGARD TO
.\" THIS SOFTWARE, INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
.\" AND FITNESS, IN NO EVENT SHALL CARNEGIE MELLON UNIVERSITY BE LIABLE
.\" FOR ANY SPECIAL, INDIRECT OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN
.\" AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING
.\" OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\" 
.\" $Id: lmtpd.8,v 1.6 2002/10/03 19:02:41 ken3 Exp $
.SH NAME
lmtpd \- LMTP server process
.SH SYNOPSIS
.B lmtpd
[
.B \-C
.I config-file
]
[
.B \-U
.I uses
]
[
.B \-T
.I timeout
]
[
.B \-D
]
.br
      [
.B \-a
]
.SH DESCRIPTION
.I Lmtpd
is an LMTP server used to deliver mail to the IMAP mailstore.
It accepts commands on its standard input and responds on its standard output.
It MUST be invoked by
.IR cyrmaster (8)
with those descriptors attached to a remote client connection.
.PP
.I Lmtpd
reads its configuration options out of the
.IR imapd.conf (5)
file unless specified otherwise by \fB-C\fR.
.SH OPTIONS
.TP
.BI \-C " config-file"
Read configuration options from \fIconfig-file\fR.
.TP
.BI \-U " uses"
The maximum number of times that the process should be used for new
connections before shutting down.  The default is 250.
.TP
.BI \-T " timeout"
The number of seconds that the process will wait for a new connection
before shutting down.  Note that a value of 0 (zero) will disable the
timeout.  The default is 60.
.TP
.BI \-D
Run external debugger specified in debug_command.
.TP
.BI \-a
Preauthorize connections initiated on an internet socket, instead of
requiring LMTP AUTH.  This should only be used for connections coming
from trusted hosts.
.SH FILES
.TP
.B /etc/imapd.conf
.SH SEE ALSO
.PP
\fBimapd.conf(5)\fR, \fBcyrmaster(8)\fR


debian/changelog
kolab-cyrus-imapd (2.2.13p1-0.4) unstable; urgency=low

  * Non-maintainer upload.
  * Fix pending l10n issues. Debconf translations:
    - Danish (Joe Hansen).  Closes: #633377
    - Dutch; (Jeroen Schot).  Closes: #656627
    - Polish (Michał Kułach).  Closes: #659569

 -- Christian Perrier <bubulle@debian.org>  Mon, 20 Feb 2012 07:15:16 +0100

kolab-cyrus-imapd (2.2.13p1-0.3) unstable; urgency=low

  * Remove db versions from debian/patches/99-update-autoconf.dpatch
    as well to fix wrong linking to db-4.7 on mipsel

 -- Ondřej Surý <ondrej@debian.org>  Tue, 28 Jun 2011 12:03:46 +0200

kolab-cyrus-imapd (2.2.13p1-0.2) unstable; urgency=low

  * Always set RET value (thanks to Steven Kurylo for catching that)

 -- Ondřej Surý <ondrej@debian.org>  Thu, 09 Jun 2011 10:07:59 +0200

kolab-cyrus-imapd (2.2.13p1-0.1) unstable; urgency=low

  * Non-maintainer upload.
    + Security upload
    + Per maintainers request
  * Imported Upstream version 2.2.13p1
  * Fix CVE-2011-1926: STARTTLS plaintext command injection
    vulnerability (VU#555316) (Closes: #629350)
  * Update autotools script before configuring by running autoreconf
  * Add support for linking with Berkeley DB5.1 (Closes: #621355)
  * Build against generic libdb-dev
  * Add automatic database upgrade script (pulled from cyrus-imapd-2.4)
    + Don't fail the upgrade if dbX.Y-util is not found
    + Try upgrade script in init.d script without upgraded databases
  * Update patches for 2.2.13p1 release
  * Filemode changes
  * Merge patches from cyrus-imapd-2.2
    + Replace sv_{yes,no,undef} with PL_sv_{yes,no,undef}
      (Closes: #628632)
    + Fix infinite loop in case of corrupted index files
    + Some GNU Hurd build fixes
    + Fix Perl IMAP.xs croak FTBFS
    + Fix cyrmaster -j (janitor) segfault
  * Build with dynamic version of com_err
  * Cherry-pick b6fd04c4 from upstream git to fix: Corrupt quota files
    cause data loss (Closes: #411240)
  * Recommend all dbX.Y-util

 -- Ondřej Surý <ondrej@debian.org>  Mon, 06 Jun 2011 21:59:10 +0200

kolab-cyrus-imapd (2.2.13-9) unstable; urgency=medium

  * Synced against cyrus-imapd package
    - Upgrade to use BerkeleyDB 4.7. (Closes: #524427)
    - Use quilt (Closes: #560628)
  * Enable imaps by default
  * Upload to unstable with urgency medium for the libkrb transition

 -- Mathieu Parent <sathieu@debian.org>  Thu, 28 Jan 2010 15:40:50 +0100

cyrus-imapd-2.2 (2.2.13-19) unstable; urgency=low

  * Switch to BerkeleyDB 4.7, the version OpenLDAP uses.
  * Add patch by Cristian Rigamonti to fix logcheck rules for "defaultbc
    doesn't exist" error message. (Closes: #511030)
  * Update Vietnamese debconf translation, thanks Clytie Siddall.
    (Closes: #548052)
  * Upload to unstable for the libkrb transition.

 -- Christoph Berg <myon@debian.org>  Sat, 23 Jan 2010 23:35:21 +0100

cyrus-imapd-2.2 (2.2.13-18) experimental; urgency=low

  [ Henrique de Moraes Holschuh ]
  * sieve/bc_eval.c (0025-upstream-fix-cve-2009-3235.dpatch):
    update for completeness to match the patch used by the security-team:
    use snprintf for scount, to future-proof against "int" larger than
    64 bits.

  [ Christoph Berg ]
  * Add myself to Uploaders.
  * Upgrade to use BerkeleyDB 4.8. (Closes: #421942)
  * Convert to use quilt, and update the patch headers to use clean paths.
    (Closes: #563303)
  * Add patch by Mathieu Parent to fix conflicting getline definition.
    (Closes: #552865)

 -- Christoph Berg <myon@debian.org>  Fri, 22 Jan 2010 22:16:02 +0100

kolab-cyrus-imapd (2.2.13-8) unstable; urgency=low

  * Remove kolab-cyrus-common.linda 
  * 01-fix_Makefile.in.dpatch: sync with cyrus-imapd-2.2
    (add '@WARNERROR@ -fPIC' to CFLAGS)
  * Disable 101-fix_Makefile.in.dpatch (redundant with above patch)
  * README.source: describe "cyrus-to-kolab" process
  * 26-local-getline.dpatch: Fix "FTBFS: ../makedepend/def.h:164: error:
     conflicting types for 'getline'" (Closes: #552848)

 -- Mathieu Parent <sathieu@debian.org>  Wed, 11 Nov 2009 23:26:07 +0100

kolab-cyrus-imapd (2.2.13-7) unstable; urgency=high

  * Setting urgency to high due to the CVE below
  * cyrus.conf: Fix "sieve port is not available form other hosts"
    (Closes: #549597)

 -- Mathieu Parent <sathieu@debian.org>  Sun, 04 Oct 2009 19:41:13 +0200

kolab-cyrus-imapd (2.2.13-6) unstable; urgency=low

  * Synced against cyrus-imapd package
  * Fix and acknowledge NMU for "CVE-2009-2632: Buffer overflow in the SIEVE
    script component" (Closes: #547712)
  * Added me as uploader
  * debian/control:
    - Prefix packages descriptions by "Kolab " 
    - Conflicts, Replaces and Provides corresponding cyrus-imapd-2.2 packages
    - Added Recommends and Suggests (based on cyrus-imapd-2.2) 
  * Get closer to cyrus-imapd-2.2: cyrus.conf 

 -- Mathieu Parent <sathieu@debian.org>  Sun, 04 Oct 2009 14:56:07 +0200

cyrus-imapd-2.2 (2.2.13-17) unstable; urgency=high

  * Security Update: CVE-2009-3235:
    Multiple stack-based buffer overflows in the Sieve parsing code,
    patches taken from upstream CVS (closes: #547947)

 -- Henrique de Moraes Holschuh <hmh@debian.org>  Tue, 22 Sep 2009 17:17:17 -0300

cyrus-imapd-2.2 (2.2.13-16) unstable; urgency=high

  * Justification for high urgency: Get security update from -15 into
    testing as fast as possible since -15 was removed before it could
    migrate.
  * Fix build dependency on tcl-dev (keeping old tcl8.3-dev to make
    backports easier) (Closes: 545281) 
  * Fix up README.Debian.database (Closes: 543293)
  * Build against newer cyrus-sasl (Closes: 527495,528468)
  * Fix cyrus-common postinst to only run makedirs if a new upstream
    is installed (Closes: 545512), add a README to note that the
    postinst needs to be changed with each new upstream
  * Fix find syntax in cyrus-makedirs (Closes: 545554)

 -- Sven Mueller <sven@debian.org>  Tue, 08 Sep 2009 10:04:14 +0200


imap/lmtpd.c
/* lmtpd.c -- Program to deliver mail to a mailbox
 *
 * $Id: lmtpd.c,v 1.144 2006/05/10 16:27:17 murch Exp $
 * Copyright (c) 1998-2003 Carnegie Mellon University.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer. 
 *
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in
 *    the documentation and/or other materials provided with the
 *    distribution.
 *
 * 3. The name "Carnegie Mellon University" must not be used to
 *    endorse or promote products derived from this software without
 *    prior written permission. For permission or any other legal
 *    details, please contact  
 *      Office of Technology Transfer
 *      Carnegie Mellon University
 *      5000 Forbes Avenue
 *      Pittsburgh, PA  15213-3890
 *      (412) 268-4387, fax: (412) 268-7395
 *      tech-transfer@andrew.cmu.edu
 *
 * 4. Redistributions of any form whatsoever must retain the following
 *    acknowledgment:
 *    "This product includes software developed by Computing Services
 *     at Carnegie Mellon University (http://www.cmu.edu/computing/)."
 *
 * CARNEGIE MELLON UNIVERSITY DISCLAIMS ALL WARRANTIES WITH REGARD TO
 * THIS SOFTWARE, INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
 * AND FITNESS, IN NO EVENT SHALL CARNEGIE MELLON UNIVERSITY BE LIABLE
 * FOR ANY SPECIAL, INDIRECT OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN
 * AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING
 * OUT OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 *
 *
 */

#include <config.h>

#ifdef HAVE_UNISTD_H
#include <unistd.h>
#endif
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <fcntl.h>
#include <sys/stat.h>
#include <syslog.h>
#include <errno.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <netdb.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <sasl/sasl.h>
#include <sasl/saslutil.h>

#include "acl.h"
#include "annotate.h"
#include "assert.h"
#include "util.h"
#include "auth.h"
#include "prot.h"
#include "imparse.h"
#include "lock.h"
#include "global.h"
#include "exitcodes.h"
#include "imap_err.h"
#include "mailbox.h"
#include "xmalloc.h"
#include "version.h"
#include "duplicate.h"
#include "append.h"
#include "mboxlist.h"
#include "notify.h"
#include "idle.h"
#include "tls.h"

#include "lmtpd.h"
#include "lmtpengine.h"
#include "lmtpstats.h"
#ifdef USE_SIEVE
#include "lmtp_sieve.h"

static sieve_interp_t *sieve_interp = NULL;
#endif

/* forward declarations */
static int deliver(message_data_t *msgdata, char *authuser,
		   struct auth_state *authstate);
static int verify_user(const char *user, const char *domain, const char *mailbox,
		       long quotacheck,
		       struct auth_state *authstate);
static char *generate_notify(message_data_t *m);

void shut_down(int code);

static FILE *spoolfile(message_data_t *msgdata);
static void removespool(message_data_t *msgdata);

/* current namespace */
static struct namespace lmtpd_namespace;

struct lmtp_func mylmtp = { &deliver, &verify_user, &shut_down,
			    &spoolfile, &removespool, &lmtpd_namespace,
			    0, 1, 0 };

static void usage();

/* global state */
const int config_need_data = CONFIG_NEED_PARTITION_DATA;

extern int optind;
extern char *optarg;
static int dupelim = 1;		/* eliminate duplicate messages with
				   same message-id */
static int singleinstance = 1;	/* attempt single instance store */
const char *BB = "";

struct stagemsg *stage = NULL;

/* per-user/session state */
static struct protstream *deliver_out, *deliver_in;
int deliver_logfd = -1; /* used in lmtpengine.c */

static struct sasl_callback mysasl_cb[] = {
    { SASL_CB_GETOPT, &mysasl_config, NULL },
    { SASL_CB_PROXY_POLICY, &mysasl_proxy_policy, NULL },
    { SASL_CB_CANON_USER, &mysasl_canon_user, NULL },
    { SASL_CB_LIST_END, NULL, NULL }
};


int service_init(int argc __attribute__((unused)), 
		 char **argv __attribute__((unused)), 
		 char **envp __attribute__((unused)))
{
    int r;

    if (geteuid() == 0) return 1;
    
    signals_set_shutdown(&shut_down);
    signal(SIGPIPE, SIG_IGN);

    singleinstance = config_getswitch(IMAPOPT_SINGLEINSTANCESTORE);
    BB = config_getstring(IMAPOPT_POSTUSER);

    global_sasl_init(0, 1, mysasl_cb);

    dupelim = config_getswitch(IMAPOPT_DUPLICATESUPPRESSION);

#ifdef USE_SIEVE
    mylmtp.addheaders = xzmalloc(2 * sizeof(struct addheader));
    mylmtp.addheaders[0].name = "X-Sieve";
    mylmtp.addheaders[0].body = SIEVE_VERSION;

    /* setup sieve support */
    sieve_interp = setup_sieve();
#else
    if (dupelim)
#endif
    {
	/* initialize duplicate delivery database */
	if (duplicate_init(NULL, 0) != 0) {
	    fatal("lmtpd: unable to init duplicate delivery database",
		  EC_SOFTWARE);
	}
    }

    /* so we can do mboxlist operations */
    mboxlist_init(0);
    mboxlist_open(NULL);

    /* so we can do quota operations */
    quotadb_init(0);
    quotadb_open(NULL);

    /* Initialize the annotatemore db (for sieve on shared mailboxes) */
    annotatemore_init(0, NULL, NULL);
    annotatemore_open(NULL);

    /* setup for sending IMAP IDLE notifications */
    idle_enabled();

    /* Set namespace */
    if ((r = mboxname_init_namespace(&lmtpd_namespace, 0)) != 0) {
	syslog(LOG_ERR, error_message(r));
	fatal(error_message(r), EC_CONFIG);
    }

    /* create connection to the SNMP listener, if available. */
    snmp_connect(); /* ignore return code */
    snmp_set_str(SERVER_NAME_VERSION, CYRUS_VERSION);

    return 0;
}

/*
 * run for each accepted connection
 */
int service_main(int argc, char **argv, 
		 char **envp __attribute__((unused)))
{
    int opt;

    deliver_in = prot_new(0, 0);
    deliver_out = prot_new(1, 1);
    prot_setflushonread(deliver_in, deliver_out);
    prot_settimeout(deliver_in, 360);

    while ((opt = getopt(argc, argv, "a")) != EOF) {
	switch(opt) {
	case 'a':
	    mylmtp.preauth = 1;
	    break;

	default:
	    usage();
	}
    }

    snmp_increment(TOTAL_CONNECTIONS, 1);
    snmp_increment(ACTIVE_CONNECTIONS, 1);

    lmtpmode(&mylmtp, deliver_in, deliver_out, 0);

    /* free session state */
    if (deliver_in) prot_free(deliver_in);
    if (deliver_out) prot_free(deliver_out);
    deliver_in = deliver_out = NULL;

    if (deliver_logfd != -1) {
	close(deliver_logfd);
	deliver_logfd = -1;
    }

    cyrus_reset_stdio(); 

    return 0;
}

/* Called by service API to shut down the service */
void service_abort(int error)
{
    shut_down(error);
}

static void
usage()
{
    fprintf(stderr, "421-4.3.0 usage: lmtpd [-C <alt_config>] [-a]\r\n");
    fprintf(stderr, "421 4.3.0 %s\n", CYRUS_VERSION);
    exit(EC_USAGE);
}

/* places msg in mailbox mailboxname.  
 * if you wish to use single instance store, pass stage as non-NULL
 * if you want to deliver message regardless of duplicates, pass id as NULL
 * if you want to notify, pass user
 * if you want to force delivery (to force delivery to INBOX, for instance)
 * pass acloverride
 */
int deliver_mailbox(struct protstream *msg,
		    struct stagemsg *stage,
		    unsigned size __attribute__((unused)),
		    char **flag,
		    int nflags,
		    char *authuser,
		    struct auth_state *authstate,
		    char *id,
		    const char *user,
		    char *notifyheader,
		    const char *mailboxname,
		    int quotaoverride,
		    int acloverride)
{
    int r;
    struct appendstate as;
    time_t now = time(NULL);
    unsigned long uid;
    const char *notifier;

    if (dupelim && id && 
	duplicate_check(id, strlen(id), mailboxname, strlen(mailboxname))) {
	/* duplicate message */
	duplicate_log(id, mailboxname, "delivery");
	return 0;
    }

    r = append_setup(&as, mailboxname, MAILBOX_FORMAT_NORMAL,
		     authuser, authstate, acloverride ? 0 : ACL_POST, 
		     quotaoverride ? -1 : 0);

    if (!r) {
	prot_rewind(msg);
	r = append_fromstage(&as, stage, now,
			     (const char **) flag, nflags, !singleinstance);
	if (r || (dupelim && id && 
	    duplicate_check(id, strlen(id), mailboxname, strlen(mailboxname)))) 
	{
	    append_abort(&as);
                    
	    if (!r) {
	      /* duplicate message */
		duplicate_log(id, mailboxname, "delivery");
		return 0;
	    }            
	} else {
	    if (dupelim && id) 
		duplicate_mark(id, strlen(id), mailboxname, 
			       strlen(mailboxname), now, uid);

	    append_commit(&as, quotaoverride ? -1 : 0, NULL, &uid, NULL);
	    syslog(LOG_INFO, "Delivered: %s to mailbox: %s", id, mailboxname);
	}
    }

    if (!r && user && (notifier = config_getstring(IMAPOPT_MAILNOTIFIER))) {
	char inbox[MAX_MAILBOX_NAME+1];
	char namebuf[MAX_MAILBOX_NAME+1];
	char userbuf[MAX_MAILBOX_NAME+1];
	const char *notify_mailbox = mailboxname;
	int r2;

	/* translate user.foo to INBOX */
	if (!(*lmtpd_namespace.mboxname_tointernal)(&lmtpd_namespace,
						    "INBOX", user, inbox)) {
	    int inboxlen = strlen(inbox);
	    if (strlen(mailboxname) >= inboxlen &&
		!strncmp(mailboxname, inbox, inboxlen) &&
		(!mailboxname[inboxlen] || mailboxname[inboxlen] == '.')) {
		strlcpy(inbox, "INBOX", sizeof(inbox)); 
		strlcat(inbox, mailboxname+inboxlen, sizeof(inbox));
		notify_mailbox = inbox;
	    }
	}

	/* translate mailboxname */
	r2 = (*lmtpd_namespace.mboxname_toexternal)(&lmtpd_namespace,
						    notify_mailbox,
						    user, namebuf);
	if (!r2) {
	    strlcpy(userbuf, user, sizeof(userbuf));
	    /* translate any separators in user */
	    mboxname_hiersep_toexternal(&lmtpd_namespace, userbuf,
					config_virtdomains ?
					strcspn(userbuf, "@") : 0);
	    notify(notifier, "MAIL", NULL, userbuf, namebuf, 0, NULL,
		   notifyheader ? notifyheader : "");
	}
    }

    return r;
}

int deliver_local(deliver_data_t *mydata, char **flag, int nflags,
		  const char *username, const char *mailboxname)
{
    char namebuf[MAX_MAILBOX_NAME+1] = "", *tail;
    message_data_t *md = mydata->m;
    int quotaoverride = msg_getrcpt_ignorequota(md, mydata->cur_rcpt);
    int ret;

    /* case 1: shared mailbox request */
    if (!*username || username[0] == '@') {
	if (*username) snprintf(namebuf, sizeof(namebuf), "%s!", username+1);
	strlcat(namebuf, mailboxname, sizeof(namebuf));

	return deliver_mailbox(md->data, mydata->stage,
			       md->size, flag, nflags,
			       mydata->authuser, mydata->authstate, md->id,
			       NULL, mydata->notifyheader,
			       namebuf, quotaoverride, 0);
    }

    /* case 2: ordinary user */
    ret = (*mydata->namespace->mboxname_tointernal)(mydata->namespace,
						    "INBOX",
						    username, namebuf);

    if (!ret) {
	ret = 1;

	tail = namebuf + strlen(namebuf);
	if (mailboxname) {
	    strlcat(namebuf, ".", sizeof(namebuf));
	    strlcat(namebuf, mailboxname, sizeof(namebuf));

	    ret = deliver_mailbox(md->data, mydata->stage,
				  md->size, flag, nflags,
				  mydata->authuser, mydata->authstate, md->id,
				  username, mydata->notifyheader,
				  namebuf, quotaoverride, 0);
	}
	if (ret) {
	    /* normal delivery to INBOX */
	    struct auth_state *authstate = auth_newstate(username);

	    *tail = '\0';

	    ret = deliver_mailbox(md->data, mydata->stage,
				  md->size, flag, nflags,
				  (char *) username, authstate, md->id,
				  username, mydata->notifyheader,
				  namebuf, quotaoverride, 1);

	    if (authstate) auth_freestate(authstate);
	}
    }

    return ret;
}

int deliver(message_data_t *msgdata, char *authuser,
	    struct auth_state *authstate)
{
    int n, nrcpts;
    char *notifyheader;
    deliver_data_t mydata;
    
    assert(msgdata);
    nrcpts = msg_getnumrcpt(msgdata);
    assert(nrcpts);

    notifyheader = generate_notify(msgdata);

    /* create 'mydata', our per-delivery data */
    mydata.m = msgdata;
    mydata.stage = stage;
    mydata.notifyheader = notifyheader;
    mydata.namespace = &lmtpd_namespace;
    mydata.authuser = authuser;
    mydata.authstate = authstate;
    
    /* loop through each recipient, attempting delivery for each */
    for (n = 0; n < nrcpts; n++) {
	char namebuf[MAX_MAILBOX_NAME+1] = "";
	char userbuf[MAX_MAILBOX_NAME+1];
	const char *user, *domain, *mailbox;
	int r = 0;

	msg_getrcpt(msgdata, n, &user, &domain, &mailbox);

	namebuf[0] = '\0';
	userbuf[0] = '\0';

	if (domain) snprintf(namebuf, sizeof(namebuf), "%s!", domain);

	/* case 1: shared mailbox request */
	if (!user) {
	    strlcat(namebuf, mailbox, sizeof(namebuf));
	}

	/* case 2: ordinary user */
	else {
	    strlcat(namebuf, "user.", sizeof(namebuf));
	    strlcat(namebuf, user, sizeof(namebuf));

	    strlcpy(userbuf, user, sizeof(userbuf));
	}
	if (domain) {
	    strlcat(userbuf, "@", sizeof(userbuf));
	    strlcat(userbuf, domain, sizeof(userbuf));
	}

	mydata.cur_rcpt = n;
#ifdef USE_SIEVE
	r = run_sieve(user, domain, mailbox, sieve_interp, &mydata);
	/* if there was no sieve script, or an error during execution,
	   r is non-zero and we'll do normal delivery */
#else
	r = 1;		/* normal delivery */
#endif

	if (r) {
	    r = deliver_local(&mydata, NULL, 0, userbuf, mailbox);
	}

	msg_setrcpt_status(msgdata, n, r);
    }

    append_removestage(stage);
    stage = NULL;
    if (notifyheader) free(notifyheader);

    return 0;
}

void fatal(const char* s, int code)
{
    static int recurse_code = 0;
    
    if(recurse_code) {
	/* We were called recursively. Just give up */
	snmp_increment(ACTIVE_CONNECTIONS, -1);
	exit(recurse_code);
    }
    recurse_code = code;
    if(deliver_out) {
	prot_printf(deliver_out,"421 4.3.0 lmtpd: %s\r\n", s);
	prot_flush(deliver_out);
    }
    if (stage) append_removestage(stage);

    syslog(LOG_ERR, "FATAL: %s", s);
    
    /* shouldn't return */
    shut_down(code);

    exit(code);
}

/*
 * Cleanly shut down and exit
 */
void shut_down(int code) __attribute__((noreturn));
void shut_down(int code)
{
#ifdef USE_SIEVE
    sieve_interp_free(&sieve_interp);
#else
    if (dupelim)
#endif
	duplicate_done();

    mboxlist_close();
    mboxlist_done();

    quotadb_close();
    quotadb_done();

    annotatemore_close();
    annotatemore_done();
#ifdef HAVE_SSL
    tls_shutdown_serverengine();
#endif
    if (deliver_out) {
	prot_flush(deliver_out);

	/* one less active connection */
	snmp_increment(ACTIVE_CONNECTIONS, -1);
    }

    cyrus_done();

    /* shutdown socket nicely */
    cyrus_reset_stdio();

    exit(code);
}

static int verify_user(const char *user, const char *domain, const char *mailbox,
		       long quotacheck, struct auth_state *authstate)
{
    char namebuf[MAX_MAILBOX_NAME+1] = "";
    int r = 0;

    if ((!user && !mailbox) ||
	(domain && (strlen(domain) + 1 > sizeof(namebuf)))) {
	r = IMAP_MAILBOX_NONEXISTENT;
    } else {
	/* construct the mailbox that we will verify */
	if (domain) snprintf(namebuf, sizeof(namebuf), "%s!", domain);

	if (!user) {
	    /* shared folder */
	    if (strlen(namebuf) + strlen(mailbox) > sizeof(namebuf)) {
		r = IMAP_MAILBOX_NONEXISTENT;
	    } else {
		strlcat(namebuf, mailbox, sizeof(namebuf));
	    }
	} else {
	    /* ordinary user -- check INBOX */
	    if (strlen(namebuf) + 5 + strlen(user) > sizeof(namebuf)) {
		r = IMAP_MAILBOX_NONEXISTENT;
	    } else {
		strlcat(namebuf, "user.", sizeof(namebuf));
		strlcat(namebuf, user, sizeof(namebuf));
	    }
	}
    }

    if (!r) {
	/*
	 * check to see if mailbox exists and we can append to it:
	 *
	 * - must have posting privileges on shared folders
	 * - don't care about ACL on INBOX (always allow post)
	 * - don't care about message size (1 msg over quota allowed)
	 */
	r = append_check(namebuf, MAILBOX_FORMAT_NORMAL, authstate,
			 !user ? ACL_POST : 0, quotacheck > 0 ? 0 : quotacheck);
    }

    if (r) syslog(LOG_DEBUG, "verify_user(%s) failed: %s", namebuf,
		  error_message(r));

    return r;
}

const char *notifyheaders[] = { "From", "Subject", "To", 0 };
/* returns a malloc'd string that should be sent to users for successful
   delivery of 'm'. */
char *generate_notify(message_data_t *m)
{
    const char **body;
    char *ret = NULL;
    unsigned int len = 0;
    unsigned int pos = 0;
    int i;

    for (i = 0; notifyheaders[i]; i++) {
	const char *h = notifyheaders[i];
	body = msg_getheader(m, h);
	if (body) {
	    int j;

	    for (j = 0; body[j] != NULL; j++) {
		/* put the header */
		/* need: length + ": " + '\0'*/
		while (pos + strlen(h) + 3 > len) {
		    ret = xrealloc(ret, len += 1024);
		}
		pos += sprintf(ret + pos, "%s: ", h);
		
		/* put the header body.
		   xxx it would be nice to linewrap.*/
		/* need: length + '\n' + '\0' */
		while (pos + strlen(body[j]) + 2 > len) {
		    ret = xrealloc(ret, len += 1024);
		}
		pos += sprintf(ret + pos, "%s\n", body[j]);
	    }
	}
    }

    return ret;
}

FILE *spoolfile(message_data_t *msgdata)
{
    int i, n;
    time_t now = time(NULL);
    FILE *f = NULL;

    /* spool to the stage of one of the recipients */
    n = msg_getnumrcpt(msgdata);
    for (i = 0; !f && (i < n); i++) {
	int r = 0;
	char namebuf[MAX_MAILBOX_NAME+1] = "";
	const char *user, *domain, *mailbox;

	/* build the mailboxname from the recipient address */
	msg_getrcpt(msgdata, i, &user, &domain, &mailbox);

	if (domain) snprintf(namebuf, sizeof(namebuf), "%s!", domain);

	/* case 1: shared mailbox request */
	if (!user) {
	    strlcat(namebuf, mailbox, sizeof(namebuf));
	}

	/* case 2: ordinary user */
	else {
	    /* assume delivery to INBOX for now */
	    strlcat(namebuf, "user.", sizeof(namebuf));
	    strlcat(namebuf, user, sizeof(namebuf));
	}
#if 0
	/* case 3: unable to handle rcpt */
	else {
	    /* force error and we'll fallback to using /tmp */
	    r = 1;
	}
#endif
	if (!r) {
	    /* setup stage for later use by deliver() */
	    f = append_newstage(namebuf, now, 0, &stage);
	}
    }

    return f;
}

void removespool(message_data_t *msgdata __attribute__((unused)))
{
    append_removestage(stage);
    stage = NULL;
}
